From fc49101f219648811ed3b599e3114e9b9b36c5f5 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:04:53 -0400
Subject: [fs] btrfs: create sprout should rename fsid on the sysfs as well

Message-id: <1407179109-13714-407-git-send-email-sandeen@redhat.com>
Patchwork-id: 87342
O-Subject: [RHEL7.1 PATCH 406/422] btrfs: create sprout should rename fsid on the sysfs as well
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Anand Jain <Anand.Jain@oracle.com>

commit b2373f255cacdc1ea4da25e75a5a78949ffd9d66
Author: Anand Jain <Anand.Jain@oracle.com>
Date:   Tue Jun 3 11:36:03 2014 +0800

    btrfs: create sprout should rename fsid on the sysfs as well

    Creating sprout will change the fsid of the mounted root.
    do the same on the sysfs as well.

    reproducer:
     mount /dev/sdb /btrfs (seed disk)
     btrfs dev add /dev/sdc /btrfs
     mount -o rw,remount /btrfs
     btrfs dev del /dev/sdb /btrfs
     mount /dev/sdb /btrfs

    Error:
    kobject_add_internal failed for fe350492-dc28-4051-a601-e017b17e6145 with -EEXIST, don't try to register things with the same name in the same directory.

    Signed-off-by: Anand Jain <anand.jain@oracle.com>
    Reviewed-by: David Sterba <dsterba@suse.cz>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/volumes.c b/fs/btrfs/volumes.c
index ccdb77e..7fed530 100644
--- a/fs/btrfs/volumes.c
+++ b/fs/btrfs/volumes.c
@@ -2154,6 +2154,7 @@ int btrfs_init_new_device(struct btrfs_root *root, char *device_path)
  mutex_unlock(&root->fs_info->fs_devices->device_list_mutex);
 
  if (seeding_dev) {
+  char fsid_buf[BTRFS_UUID_UNPARSED_SIZE];
   ret = init_first_rw_device(trans, root, device);
   if (ret) {
    btrfs_abort_transaction(trans, root, ret);
@@ -2164,6 +2165,14 @@ int btrfs_init_new_device(struct btrfs_root *root, char *device_path)
    btrfs_abort_transaction(trans, root, ret);
    goto error_trans;
   }
+
+  /* Sprouting would change fsid of the mounted root,
+   * so rename the fsid on the sysfs
+   */
+  snprintf(fsid_buf, BTRFS_UUID_UNPARSED_SIZE, "%pU",
+      root->fs_info->fsid);
+  if (kobject_rename(&root->fs_info->super_kobj, fsid_buf))
+   goto error_trans;
  } else {
   ret = btrfs_add_device(trans, root, device);
   if (ret) {
diff --git a/include/uapi/linux/btrfs.h b/include/uapi/linux/btrfs.h
index 6f9c38c..2f47824 100644
--- a/include/uapi/linux/btrfs.h
+++ b/include/uapi/linux/btrfs.h
@@ -38,6 +38,7 @@ struct btrfs_ioctl_vol_args {
 #define BTRFS_SUBVOL_QGROUP_INHERIT (1ULL << 2)
 #define BTRFS_FSID_SIZE 16
 #define BTRFS_UUID_SIZE 16
+#define BTRFS_UUID_UNPARSED_SIZE 37
 
 #define BTRFS_QGROUP_INHERIT_SET_LIMITS (1ULL << 0)
 
-- 
1.7.1